subsys/net/lib/sockets/sockets_tls.c

```diff
@@ -1711,58 +1776,117 @@ static int ztls_poll_update_ctx(struct tls_context *ctx,
 	}
 
 	if (pfd->events & ZSOCK_POLLIN) {
-		if (!ctx->is_listening) {
-			/* Already had TLS data to read on socket. */
-			if (mbedtls_ssl_get_bytes_avail(&ctx->ssl) > 0) {
-				pfd->revents |= ZSOCK_POLLIN;
-				goto next;
-			}
+		ret = ztls_poll_update_pollin(pfd->fd, ctx, pfd);
+		if (ret == -EAGAIN && pfd->revents != 0) {
+			(*pev - 1)->state = K_POLL_STATE_NOT_READY;
+			return -EAGAIN;
 		}
+	}
 
-		if (!(pfd->revents & ZSOCK_POLLIN)) {
-			/* No new data on a socket. */
-			goto next;
+	return ret;
+}
+
+static inline int ztls_poll_offload(struct pollfd *fds, int nfds, int timeout)
+{
+	int fd_backup[CONFIG_NET_SOCKETS_POLL_MAX];
+	const struct fd_op_vtable *vtable;
+	void *ctx;
+	int ret = 0;
+	int result;
+	bool retry;
+	int remaining;
+	uint32_t entry = k_uptime_get_32();
+
+	/* Overwrite TLS file decriptors with underlying ones. */
+	for (int i = 0; i < nfds; i++) {
```

Comment:
There are several `for (int i = ...` settings in this function. I suggest declaring the `i` above so that we can use it in all invocations.